# Natural selection and the distribution of chromosomal inversion lengths

## Overview

This is a GitHub repository hosting the supplementary R code for the a theoretical population genetics research project that is now in review under the title "*Natural selection and the distribution of chromosomal inversion lengths*" (doi: XXXXXX). Here you can find all of the necessary R code to reproduce the key simulation results presented in the published paper and appendices. Other supplementary material for the paper can be downloaded as an online appendix from [link](URL).


## Abstract

Chromosomal inversions contribute substantially to genome evolution, yet the processes governing their fixation remain poorly understood. Theory suggests that a readily measurable property of inversions—their length—can potentially affect their evolutionary fates. Emerging data on the lengths of fixed inversions may therefore provide clues to the evolutionary processes underlying inversion substitutions. However, formal theoretical predictions for the distribution of inversion lengths remain incomplete, making empirical patterns difficult to interpret. We model the relation between inversion length and fixation probability for neutral inversions, underdominant inversions, directly beneficial inversions (e.g., those with favorable breakpoints), and indirectly beneficial inversions that capture locally adapted alleles at migration-selection balance and suppress recombination between them. We also consider how segregating deleterious mutations affect the lengths of inversion substitutions. When neutral, underdominant, or directly beneficial, small inversions contribute the most to genome divergence. In contrast, inversions of intermediate length contribute disproportionately to divergence under the local adaptation scenario. These predictions are robust to effects of deleterious genetic variation, which shift inversion length distributions towards smaller sizes without fundamentally altering their qualitative forms. We also find that the four categories of inversion substitution differ in sensitivity to effective population size (N~e~). The lengths of neutral inversion substitutions increase, while underdominant substitutions decrease, with N~e~; the lengths of directly and indirectly beneficial inversion substitutions are relatively insensitive to N~e~. Our models clarify how inversion length distributions relate to processes of inversion fixation, providing a platform for testing how natural selection shapes the evolution of genome structure.



## Citing information

Citing information for the resulting paper will be provided when it is made [available through the publisher](URL). You can also contact me directly if you would like a reprint. 


## Contact & bug reporting

Please report any bugs, problems, or issues by opening an issue on the inversionSize github [issues page](https://github.com/colin-olito/inversionSize/issues). If you would like to report a bug/issue, and do not have a github account (and don't want to get one), please send a brief email detailing the problem you encountered to colin.olito at biol dot lu dot se.